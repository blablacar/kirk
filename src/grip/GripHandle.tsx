import styled from 'styled-components'

export const GripHandle = styled.div`
  & {
    display: flex;
    justify-content: center;
    padding: 10px 0;
  }
  &::before {
    content: ' ';
    display: block;
    width: 44px;
    height: 4px;
    background-color: #ddd;
    border-radius: 100px;
  }
`
