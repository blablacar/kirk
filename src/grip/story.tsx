import React from 'react'
import { action } from '@storybook/addon-actions'
import { withKnobs } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'

import { ItemRadio } from '../itemRadio'
import { ItemRadioGroup } from '../itemRadioGroup'
import { TheVoice } from '../theVoice'
import { Grip } from './index'
import readme from './specifications/grip.md'

const stories = storiesOf('Widgets|Grip', module).addDecorator(withKnobs)

stories.add(
  'default',
  () => (
    <Grip onSlideUp={action('Slide Up')} onSlideDown={action('Slide Down')}>
      <TheVoice>What is your route?</TheVoice>
      <ItemRadioGroup>
        <ItemRadio labelTitle="Via A10" label="Tolls" name="foo" value="1" />
        <ItemRadio labelTitle="Via N200" label="No tolls" name="foo" value="2" />
      </ItemRadioGroup>
    </Grip>
  ),
  {
    readme: { content: readme },
  },
)
