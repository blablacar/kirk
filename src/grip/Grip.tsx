import React, { useEffect, useRef } from 'react'

import { A11yProps, pickA11yProps } from '../_utils/interfaces'
import { GripHandle } from './GripHandle'

export interface GripProps extends A11yProps {
  children?: React.ReactNode
  onSlideUp: () => void
  onSlideDown: () => void
  className?: string
  disabled?: boolean
}

export const SLIDE_OFFSET = 20 // To get more precise with feeling/testing

export const touchEndListener = (
  clientY: number,
  fingerYPosition: React.MutableRefObject<number>,
  resetFingerYPosition: () => void,
  props: GripProps,
) => {
  const { onSlideDown, onSlideUp } = props
  if (fingerYPosition !== null) {
    if (clientY < fingerYPosition.current - SLIDE_OFFSET) {
      onSlideUp()
    } else if (clientY > fingerYPosition.current + SLIDE_OFFSET) {
      onSlideDown()
    }
    resetFingerYPosition()
  }
}

export const Grip = (props: GripProps): JSX.Element => {
  const { onSlideUp, onSlideDown, children = null, className = '', disabled = false } = props
  const a11yAttrs = pickA11yProps(props)
  const fingerYPosition = useRef<number | null>(null)
  const resetFingerYPosition = () => {
    fingerYPosition.current = null
  }

  // Listening to finger being lift from the screen and check position
  // difference with starting point
  useEffect(() => {
    if (!disabled) {
      const delegatedTouchEndListener = (e: TouchEvent) =>
        touchEndListener(
          e.changedTouches.item(0).clientY,
          fingerYPosition,
          resetFingerYPosition,
          props,
        )
      window.addEventListener('touchend', delegatedTouchEndListener)
      return () => {
        window.removeEventListener('touchend', delegatedTouchEndListener)
      }
    }
    return () => {}
  }, [disabled, onSlideUp, onSlideDown])

  return (
    <div
      className={className}
      onTouchStart={e => {
        fingerYPosition.current = e.touches.item(0).clientY
      }}
      {...a11yAttrs}
    >
      <GripHandle aria-hidden="true" />
      {children}
    </div>
  )
}
