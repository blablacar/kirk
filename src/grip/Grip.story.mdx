import { action } from '@storybook/addon-actions'
import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs/blocks'

import { ItemRadio } from '../itemRadio'
import { ItemRadioGroup } from '../itemRadioGroup'
import { TheVoice } from '../theVoice'
import { Grip } from './index'

<Meta title="Widgets/Grip" />

# Grip

<Canvas>
  <Story name="Default">
    <Grip onSlideUp={action('Slide Up')} onSlideDown={action('Slide Down')}>
      <TheVoice>What is your route?</TheVoice>
      <ItemRadioGroup>
        <ItemRadio labelTitle="Via A10" label="Tolls" name="foo" value="1" />
        <ItemRadio labelTitle="Via N200" label="No tolls" name="foo" value="2" />
      </ItemRadioGroup>
    </Grip>
  </Story>
</Canvas>


## Specifications

Use it to indicate that a part of the screen is reactive to slide touch events.

```jsx
import Grip from '@blablacar/ui-library/build/grip'
<Grip
  onSlideUp={() => {
    /* ... */
  }}
  onSlideDown={() => {
    /* ... */
  }}
>
  <TheVoice>What is your route?</TheVoice>
</Grip>
```

<ArgsTable of={Grip} />
