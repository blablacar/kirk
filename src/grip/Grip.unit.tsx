import React from 'react'
import { shallow } from 'enzyme'

import { TheVoice } from '../theVoice'
import { Grip, SLIDE_OFFSET, touchEndListener } from './Grip'
import { GripHandle } from './GripHandle'

describe('Grip', () => {
  const defaultProps = {
    onSlideUp: jest.fn(),
    onSlideDown: jest.fn(),
  }
  it('Should render a GripHandle', () => {
    const grip = shallow(<Grip {...defaultProps} />)
    expect(grip.find(GripHandle).exists()).toBe(true)
  })
  it('Should render its children', () => {
    const grip = shallow(
      <Grip {...defaultProps}>
        <TheVoice>Hello World</TheVoice>
      </Grip>,
    )
    expect(grip.find(TheVoice).exists()).toBe(true)
  })
  describe('touchEndListener', () => {
    const fingerYPosition = 100
    const setFingerYPosition = jest.fn()
    it('Should call onSlideUp when touch position is higher than initial touch', () => {
      const slideUpMock = jest.fn()
      touchEndListener(fingerYPosition - SLIDE_OFFSET * 2, fingerYPosition, setFingerYPosition, {
        ...defaultProps,
        onSlideUp: slideUpMock,
      })
      expect(slideUpMock).toHaveBeenCalled()
    })
    it('Should call onSlideDown when touch position is lower than initial touch', () => {
      const slideDownMock = jest.fn()
      touchEndListener(fingerYPosition + SLIDE_OFFSET * 2, fingerYPosition, setFingerYPosition, {
        ...defaultProps,
        onSlideDown: slideDownMock,
      })
      expect(slideDownMock).toHaveBeenCalled()
    })
    it('Should not do anything if touch position is within the SLIDE_OFFSET', () => {
      const slideDownMock = jest.fn()
      const slideUpMock = jest.fn()
      touchEndListener(fingerYPosition - SLIDE_OFFSET / 2, fingerYPosition, setFingerYPosition, {
        onSlideUp: slideUpMock,
        onSlideDown: slideDownMock,
      })
      touchEndListener(fingerYPosition + SLIDE_OFFSET / 2, fingerYPosition, setFingerYPosition, {
        onSlideUp: slideUpMock,
        onSlideDown: slideDownMock,
      })
      expect(slideDownMock).not.toHaveBeenCalled()
      expect(slideUpMock).not.toHaveBeenCalled()
    })
    it('Should not do anything if disabled prop is set to true', () => {
      const slideDownMock = jest.fn()
      const slideUpMock = jest.fn()
      touchEndListener(fingerYPosition - SLIDE_OFFSET * 2, fingerYPosition, setFingerYPosition, {
        onSlideUp: slideUpMock,
        onSlideDown: slideDownMock,
        disabled: true,
      })
      touchEndListener(fingerYPosition + SLIDE_OFFSET * 2, fingerYPosition, setFingerYPosition, {
        onSlideUp: slideUpMock,
        onSlideDown: slideDownMock,
        disabled: true,
      })
      expect(slideDownMock).not.toHaveBeenCalled()
      expect(slideUpMock).not.toHaveBeenCalled()
    })
  })
})
