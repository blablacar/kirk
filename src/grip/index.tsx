import styled from 'styled-components'

import { Grip } from './Grip'

const StyledGrip = styled(Grip)``

export { GripProps } from './Grip'
export { StyledGrip as Grip }
export default StyledGrip
