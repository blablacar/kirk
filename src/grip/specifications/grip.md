# Grip

Use it to indicate that a part of the screen is reactive to slide touch events.

<!-- STORY -->

## Usage

`TextComponents` are exported individually.

```js
import Grip from '@blablacar/ui-library/build/grip'
;<Grip
  onSlideUp={() => {
    /* ... */
  }}
  onSlideDown={() => {
    /* ... */
  }}
>
  <TheVoice>What is your route?</TheVoice>
</Grip>
```

## Props

| **Props**     | **Value**  | **Description**                     |
| ------------- | ---------- | ----------------------------------- |
| `onSlideUp`   | `function` | Callback for slide up touch event   |
| `onSlideDown` | `function` | Callback for slide down touch event |
| `disabled`    | `boolean`  | Disables touch events               |
