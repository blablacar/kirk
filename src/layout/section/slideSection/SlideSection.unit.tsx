import React from 'react'
import { mount, shallow } from 'enzyme'

import { render } from '@testing-library/react'

import { Grip } from '../../../grip'
import { SlideSection } from './index'

describe('SlideSection', () => {
  it('Should render a Grip component', () => {
    const container = shallow(
      <SlideSection media={<div>Media</div>}>{() => <div>Children</div>}</SlideSection>,
    )
    expect(container.find(Grip).exists()).toBe(true)
  })
  it('Should render the elements from media prop', () => {
    const container = mount(
      <SlideSection media={<div>Media</div>}>{() => <div>Children</div>}</SlideSection>,
    )
    expect(container.text()).toContain('Media')
  })
  it('Should render the elements from reducedContent prop', () => {
    const container = mount(
      <SlideSection media={<div>Media</div>} reducedContent={<div>ReducedContent</div>}>
        {() => <div>Children</div>}
      </SlideSection>,
    )
    expect(container.text()).toContain('ReducedContent')
  })
  it('Should render its children', () => {
    const container = mount(
      <SlideSection media={<div>Media</div>}>{() => <div>Children</div>}</SlideSection>,
    )
    expect(container.text()).toContain('Children')
  })
  it('Should change the position when calling the children methods', () => {
    const { container, getByText } = render(
      <SlideSection media={<div>Media</div>}>
        {(setDefaultPosition, setReducedPosition, setExpandedPosition) => (
          <div>
            <button onClick={setDefaultPosition} type="button">
              default
            </button>
            <button onClick={setReducedPosition} type="button">
              reduce
            </button>
            <button onClick={setExpandedPosition} type="button">
              expand
            </button>
          </div>
        )}
      </SlideSection>,
    )
    getByText('reduce').click()
    expect(container.querySelector('.reduced')).toMatchSnapshot()

    getByText('expand').click()
    expect(container.querySelector('.expanded')).toMatchSnapshot()

    getByText('default').click()
    expect(container.querySelector('.default')).toMatchSnapshot()
  })
})
