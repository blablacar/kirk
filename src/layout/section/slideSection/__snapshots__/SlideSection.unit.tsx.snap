// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlideSection Should change the position when calling the children methods 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 24px;
}

.c1::before {
  content: ' ';
  display: block;
  width: 44px;
  height: 4px;
  background-color: #DDD;
  border-radius: 100px;
}

.c0 {
  background-color: #fff;
  height: 0px;
  position: absolute;
  top: 100%;
  left: 0;
  right: 0;
  -webkit-transition: -webkit-transform ease 300ms;
  -webkit-transition: transform ease 300ms;
  transition: transform ease 300ms;
  box-shadow: 0 4px 4px rgba(0,0,0,.25);
}

.c0.default {
  -webkit-transform: translateY(-0px);
  -ms-transform: translateY(-0px);
  transform: translateY(-0px);
}

.c0.expanded {
  -webkit-transform: translateY(-0px);
  -ms-transform: translateY(-0px);
  transform: translateY(-0px);
  overflow: scroll;
}

.c0.reduced {
  -webkit-transform: translateY(-24px);
  -ms-transform: translateY(-24px);
  transform: translateY(-24px);
}

<div
  class="reduced c0"
>
  <div
    class=""
  >
    <div
      aria-hidden="true"
      class="c1"
    />
    <div />
    <div>
      <button
        type="button"
      >
        default
      </button>
      <button
        type="button"
      >
        reduce
      </button>
      <button
        type="button"
      >
        expand
      </button>
    </div>
  </div>
</div>
`;

exports[`SlideSection Should change the position when calling the children methods 2`] = `
<div
  class="expanded sc-htpNat fxlGzn"
>
  <div
    class=""
  >
    <div
      aria-hidden="true"
      class="sc-bdVaJa lmxsFm"
    />
    <div />
    <div>
      <button
        type="button"
      >
        default
      </button>
      <button
        type="button"
      >
        reduce
      </button>
      <button
        type="button"
      >
        expand
      </button>
    </div>
  </div>
</div>
`;

exports[`SlideSection Should change the position when calling the children methods 3`] = `
<div
  class="default sc-htpNat fxlGzn"
>
  <div
    class=""
  >
    <div
      aria-hidden="true"
      class="sc-bdVaJa lmxsFm"
    />
    <div />
    <div>
      <button
        type="button"
      >
        default
      </button>
      <button
        type="button"
      >
        reduce
      </button>
      <button
        type="button"
      >
        expand
      </button>
    </div>
  </div>
</div>
`;
